clang/lib/CodeGen/Targets/SystemZ.cpp

```diff
@@ -529,9 +530,355 @@ bool SystemZTargetCodeGenInfo::isVectorTypeBased(const Type *Ty,
+            RetTy = FTSingleTy;
+            break;
+          } else if (elemKind == BT->getKind()) {
+            break;
+          } else {
+            return std::nullopt;
+          }
+        default:
+          return std::nullopt;
+        }
+      } else {
+        return std::nullopt;
+      }
+
+      Count++;
+    }
+    if (Count == 2) {
+      // The last thing that needs to be checked is the size of the struct.
+      // If we have to emit any padding (eg. because of attribute aligned), this
+      // disqualifies the type from being complex.
+      unsigned RecordSize = getContext().getTypeSize(RT);
+      unsigned ElemSize = getContext().getTypeSize(RetTy);
+      if (RecordSize > 2 * ElemSize)
+        return std::nullopt;
+      return RetTy;
+    }
+  }
+  return std::nullopt;
+}
+
+ABIArgInfo ZOSXPLinkABIInfo::classifyReturnType(QualType RetTy) const {
+
+  // Ignore void types.
+  if (RetTy->isVoidType())
+    return ABIArgInfo::getIgnore();
+
+  // Vectors are returned directly.
+  if (isVectorArgumentType(RetTy))
+    return ABIArgInfo::getDirect();
+
+  // Complex types are returned by value as per the XPLINK docs.
+  // Their members will be placed in FPRs.
+  if (RetTy->isAnyComplexType())
+    return ABIArgInfo::getDirect();
+
+  // Complex LIKE structures are returned by value as per the XPLINK docs.
+  // Their members will be placed in FPRs.
+  if (RetTy->getAs<RecordType>()) {
+    if (auto CompTy = getFPTypeOfComplexLikeType(RetTy)) {
+      llvm::Type *FPTy = CGT.ConvertType(*CompTy);
+      llvm::Type *CoerceTy = llvm::StructType::get(FPTy, FPTy);
+      auto AI = ABIArgInfo::getDirect(CoerceTy);
+      AI.setCanBeFlattened(false);
+      return AI;
+    }
+  }
+
+  // Aggregates with a size of less than 3 GPRs are returned in GRPs 1, 2 and 3.
+  // Other aggregates are passed in memory as an implicit first parameter.
+  if (isAggregateTypeForABI(RetTy)) {
+    uint64_t AggregateTypeSize = getContext().getTypeSize(RetTy);
+
+    if (AggregateTypeSize <= 3 * GPRBits) {
+      uint64_t NumElements =
+          AggregateTypeSize / GPRBits + (AggregateTypeSize % GPRBits != 0);
+
+      // Types up to 8 bytes are passed as an integer type in GPR1.
+      // Types between 8 and 16 bytes are passed as integer types in GPR1, 2.
+      // Types between 16 and 24 bytes are passed as integer types in GPR1, 2
+      // and 3.
+      llvm::Type *CoerceTy = llvm::IntegerType::get(getVMContext(), GPRBits);
+      if (NumElements > 1)
+        CoerceTy = llvm::ArrayType::get(CoerceTy, NumElements);
+      return ABIArgInfo::getDirectInReg(CoerceTy);
+    }
+    return getNaturalAlignIndirect(RetTy);
+  }
+
+  return (isPromotableIntegerType(RetTy) ? ABIArgInfo::getExtend(RetTy)
+                                         : ABIArgInfo::getDirect());
+}
+
+ABIArgInfo ZOSXPLinkABIInfo::classifyArgumentType(QualType Ty,
+                                                  bool IsNamedArg) const {
```

Comment:
The ELF version now has
```
  Ty = useFirstFieldIfTransparentUnion(Ty);
```
here.  Do we need that for XPLINK as well?

Reply:
This is about the GCC extension `__attribute__((transparent_union))` which is documented to have this semantics:

>[...] the argument is passed to the function using the calling conventions of the first member of the transparent union, not the calling conventions of the union itself. 

So I believe this would be required for all ABIs.

(Actually, I see that the SystemZ ELF code is missing this as well - it should probably be added there as well.  However, this may not make any difference in practice, as the ELF calling conventions would agree anyway.  In any case, I'll have a look into adding this on the ELF side.)